pkg/api/norman/customization/gke/handler.go

```diff
@@ -0,0 +1,194 @@
+			util.ReturnHTTPError(writer, req, errCode, err.Error())
+			return
+		}
+		writer.Write(serialized)
+	case "gkeSubnetworks":
+		if serialized, errCode, err = listSubnetworks(req.Context(), capa); err != nil {
+			logrus.Debugf("[gke-handler] error getting subnetworks: %v", err)
+			util.ReturnHTTPError(writer, req, errCode, err.Error())
+			return
+		}
+		writer.Write(serialized)
+	case "gkeVersions":
+		if serialized, errCode, err = listVersions(req.Context(), capa); err != nil {
+			logrus.Debugf("[gke-handler] error getting versions: %v", err)
+			util.ReturnHTTPError(writer, req, errCode, err.Error())
+			return
+		}
+		writer.Write(serialized)
+	case "gkeZones":
+		if serialized, errCode, err = listZones(req.Context(), capa); err != nil {
+			logrus.Debugf("[gke-handler] error getting zones: %v", err)
+			util.ReturnHTTPError(writer, req, errCode, err.Error())
+			return
+		}
+		writer.Write(serialized)
+	default:
+		util.ReturnHTTPError(writer, req, httperror.NotFound.Status, "invalid endpoint "+resourceType)
+	}
+}
+
+func (h *handler) getCloudCredential(req *http.Request, cap *Capabilities, credID string) error {
+	ns, name := ref.Parse(credID)
+	if ns == "" || name == "" {
+		logrus.Debugf("[GKE] invalid cloud credential ID %s", credID)
+		return fmt.Errorf("invalid cloud credential ID %s", credID)
+	}
+
+	var accessCred client.CloudCredential //var to check access
+	if err := access.ByID(h.generateAPIContext(req), &schema.Version, client.CloudCredentialType, credID, &accessCred); err != nil {
+		if apiError, ok := err.(*httperror.APIError); ok {
+			if apiError.Code.Status == httperror.PermissionDenied.Status || apiError.Code.Status == httperror.NotFound.Status {
+				return fmt.Errorf("cloud credential not found")
+			}
+		}
+		return err
+	}
+
+	cc, err := h.secretsLister.Get(namespace.GlobalNamespace, name)
+	if err != nil {
+		logrus.Debugf("[GKE] error accessing cloud credential %s", credID)
+		return fmt.Errorf("error accessing cloud credential %s", credID)
+	}
+	cap.Credentials = string(cc.Data["googlecredentialConfig-authEncodedJson"])
+	region := req.URL.Query().Get("region")
+	if region != "" {
+		cap.Region = region
+	}
+	zone := req.URL.Query().Get("zone")
+	if zone != "" {
+		cap.Zone = zone
+	}
+	projectID := req.URL.Query().Get("projectId")
+	if projectID != "" {
+		cap.ProjectID = projectID
+	}
+
+	return nil
+}
+
+func (h *handler) getCredentialsFromBody(writer http.ResponseWriter, req *http.Request, cap *Capabilities) error {
+	raw, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("cannot read request body: " + err.Error())
+	}
+
+	if err = json.Unmarshal(raw, &cap); err != nil {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("cannot parse request body: " + err.Error())
+	}
+
+	if cap.Region == "" {
+		writer.WriteHeader(http.StatusBadRequest)
+		return fmt.Errorf("invalid region name")
```

Comment:
I think this is only required if querying `/meta/gkeSubnetworks` https://github.com/rancher/rancher/pull/31759/files#diff-68792714fef67a0819b8cccf08dbf2e76502031d314c37c390435358d240220aL31 so probably best to leave the validation in the `listSubnetworks` function

Reply:
I moved it to proper functions, thank you for pointing it